### ISSUE TITLE
Port BSC logic to Erigon

### ISSUE THREAD

**Issue by tsutsu** [binance-chain/bsc#568]
# Rationale

Erigon (previously "turbo-geth") is a fork of go-ethereum which pursues a more-efficient and scalable implementation while retaining an ETH-1.0 design. These optimizations have been done using several techniques, such as:

* storing data un-merkled in a flat key-value store (MDBX) with versioned keys (resulting in an order-of-magnitude lower archive-node disk utilization, and so an order-of-magnitude more useful hot data in the page cache)
* running catch-up cycles where each "stage" of sync occurs for many blocks in parallel whenever possible (the blocks download in parallel, then the blocks execute as a serial batch, then the state is committed in parallel, then the receipts et al are recorded in parallel, etc.)
* separating the p2p-sync and RPC-daemon processes, with the RPC-daemon able to be operationally tuned and scaled separately to the p2p-node, or even to be run on separate hardware (presuming a network filesystem).

These features seem to be exactly what BSC needs to increase its scalability over the short term (while waiting for a long-term, ETH2-like solution.)

Erigon has been key for us (@covalenthq) in keeping the OpEx of our internal eth-mainnet RPC archive nodes to a minimum, while scaling our RPS by several orders of magnitude. Our RPC archive nodes for networks built on go-ethereum (such as BSC, but also many of BSC's "competitors") consume much of our SREs' time, continuously generating unplanned work for them. In contrast, our Erigon nodes can basically be forgotten about. Erigon just does its job, predictably and robustly.

In our ideal world, BSC (and other blockchain-node software as well!) would be pieces of Erigon-based node software, rather than pieces of geth-based node software. I think this port could serve the needs of many of BSC's users and validators.

# Implementation

We (@covalenthq) have already begun—and have seemingly come close to completing—a port of the BSC "patch set", translating any bsc-specific code from go-ethereum's architecture over to Erigon's.

The active Erigon PR can be found here: https://github.com/ledgerwatch/erigon/pull/2991

Our PR is still a WIP, because it isn't yet syncing to the tip of bsc-mainnet. It does make progress for a very short while, until block 363, when (IIRC) the first validator-reward "system transaction" occurs. (It currently stalls out due to a nonce mismatch, likely due to TXs not being executed in the order they should be, or not having the side-effects they should when executed.)

We would love for the BSC dev-team to take a look at the linked Erigon PR, and to offer any guidance you might have on getting it working. I'm sure the problem that's a current roadblock for us would be pretty trivial to solve with a deep understanding of BSC's semantics re: system-tx processing.

We hope this code can eventually be merged into upstream Erigon. Binance and Erigon _could_ work in cooperation to maintain this code _within_ upstream Erigon, rather than "bsc" being its own piece of software. bsc-mainnet is exactly the kind of network that would best stress-test and "prove out" Erigon's optimizations, so I expect that a very useful relationship could be formed here, with new Erigon optimizations being developed in light of bsc-mainnet scaling bottlenecks.

**Comment by ivan-liljeqvist** [comment]
This needs to happen!! BSC devs please help covalent here

**Comment by perwpqwe** [comment]
fantastic, that's what we need right now.

**Comment by gituser** [comment]
Erigon as the ETH client isn't mature yet enough, it still needs more time imo - just take a look at their https://github.com/ledgerwatch/erigon/issues.

Most likely Erigon's DB will require even more storage (judging from this  https://github.com/ledgerwatch/erigon/issues/2569 as pruned node takes much more space compared to geth).

On the other hand BSC devs said long time ago that they will supply Parity / OpenEthereum client for BSC chain.

**Comment by koen84** [comment]
> Erigon as the ETH client isn't mature yet enough, it still needs more time imo - just take a look at their https://github.com/ledgerwatch/erigon/issues.
> 
> Most likely Erigon's DB will require even more storage (judging from this [ledgerwatch/erigon#2569](https://github.com/ledgerwatch/erigon/issues/2569) as pruned node takes much more space compared to geth).
> 
> On the other hand BSC devs said long time ago that they will supply Parity / OpenEthereum client for BSC chain.

You're not reading that issue correctly.  On ETH mainnet, erigon is less than 2TB versus geth / openethereum around 9TB.
If ported correctly, erigon would very likely bring a LOT of space savings and also no more database compacting, as well as much faster sync times (including archive node).

**Comment by perfectcircle2020** [comment]
@gituser Also because erigon is by default in **archive mode** the occupied space you're talking about referes to the entire archive node

**Comment by gituser** [comment]
@koen84 @perfectcirle2020 no, you're both wrong. 

User in that issue ran the erigon with options `--prune=hrtc` which means it is a pruned node just to *90K last blocks limit*.
It's NOT an archive node, but it still occupied 2.5x space compared to geth.

**Comment by koen84** [comment]
@gituser we're both stating factually 100% correct information, trying to help the community understand.

I've been working with erigon (turbogeth in the day) software and their team since Summer 2020.  Have you even ran it before disparging it ?  Erigon-BSC will be a massive improvement, especially for those who need to run archive nodes.

Yes, issues and bugs are being found, you also see rapid response from the erigon team fixing them and pushing in new releases.

**Comment by KeefeL** [comment]
The full nodes of erigon is available, but not production ready yet, feel free to have a try: https://github.com/binance-chain/bsc-erigon

**Comment by perfectcircle2020** [comment]
@KeefeL what are the neccesary args to pass to the executable to run it on mainnet?

**Comment by havsar** [comment]
`./build/bin/erigon --datadir ./chain --chain bsc`

Make sure to build from the bsc branch beforehand, not main.

**Comment by koen84** [comment]
> The full nodes of erigon is available, but not production ready yet, feel free to have a try: https://github.com/binance-chain/bsc-erigon

Are you saying it cannot be used for archive nodes ?

**Comment by forcodedancing** [comment]
@koen84 Erigon can also be used to archive nodes, we are doing tests now.

**Comment by RumeelHussainbnb** [comment]
Please proceed to join our Discord channel for more discussion at https://discord.com/invite/binancesmartchain

**Comment by koen84** [comment]
@forcodedancing yeah, i have erigon-bsc archive running, it completed the 16 stages once, but now it's crash-looping in stage 6.
